### PR TITLE
Speed up reading floating-point numbers from ENDF

### PR DESCRIPTION
**Pull request by paulromano**
I recently learned that `2. 3 - 1` is a valid floating-point number in ENDF (meaning 0.23), i.e., whitespace is basically supposed to be ignored. To fix this in our current version of `float_endf` would make it even slower, and I found profiling our `IncidentNeutron.from_endf` method reveals that a bulk of the time is already spent in `float_endf`. So, what I decided to do was to write a highly-optimized version in C that handles all these weird cases along with a thin Cython layer that handles the conversion from a Python `str` to a `char*` and then from a C `double` to a Python `float`. Profiling also revealed that we were spending quite a bit of time constructing Legendre polynomial objects (which were only used if you actually evaluate the Legendre polynomial), so I changed it so that they're only created when needed. Altogether, I'm finding that `IncidentNeutron.from_endf` is now 5.5x faster on average (reading all files from ENDF/B-VII.1).

### CODE REVIEW

File: openmc/data/endf.c [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+  char arr[12]; // 11 characters plus a null terminator
+  int j = 0; // current position in arr
+  int found_decimal = 0;
+  int found_exponent = 0;
+  for (int i = 0; i < n; ++i) {
+    // Skip whitespace characters
+    char c = buffer[i];
+    if (c == ' ') continue;
+
+    if (found_decimal) {
```
Comment (smharper) [inline openmc/data/endf.c]: What about exponentials without decimals?  I think this would miss e.g. `1+3`

**Comment by smharper** [comment]
That is a disgustingly formatted float!  Out of curiosity, did you find any of these nastily formatted floats in the wild?

File: openmc/data/endf.c (hunk 1 above)
Comment (paulromano) [reply openmc/data/endf.c]: Ah, you're right. Thanks for catching that!

**Comment by paulromano** [comment]
@smharper The ones with spaces embedded, no. But I was surprised to find out that technically it is permissible. I just updated `cfloat_endf` to handle decimal-less numbers. I also decided to support numbers like "1.0d2". Although ENDF does not permit this, it has been known to occur in files (for example, the entire atomic relaxation and electron sublibraries in ENDF/B-VIII.0) so I figured we might as well have a hack to support it.

**Comment by paulromano** [comment]
@smharper Any further requests on this PR?

**Pull request merged**